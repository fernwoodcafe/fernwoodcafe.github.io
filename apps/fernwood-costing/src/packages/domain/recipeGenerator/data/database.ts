import type { AvailableCustomerOptions, CostingData, PackagingOption, PricingOptions } from "./schema";

const costForPackaging = new Map<PackagingOption, number>([
  ["cup @ 8 oz", 0.13],
  ["cup @ 12 oz", 0.17],
  ["cup @ 16 oz", 0.19],
  ["lid @ 8 oz", 0.12],
  ["lid @ 10 to 20 oz", 0.13],
  ["sleeve @ 12 to 20 oz", 0.08],
]);

const packagingForHotDrinkSizeInOunces = new Map<number, PackagingOption[]>([
  [8, ["cup @ 8 oz", "lid @ 8 oz"]],
  [12, ["cup @ 12 oz", "lid @ 10 to 20 oz", "sleeve @ 12 to 20 oz"]],
  [16, ["cup @ 16 oz", "lid @ 10 to 20 oz", "sleeve @ 12 to 20 oz"]],
]);

const customerOptions: AvailableCustomerOptions = {
  availableSizesInOunces: [12],
  availableExpressoShots: [2],
  availableMilkAlternatives: ["dairy_3_percent"],
  availableCups: ["to_go", "own_cup", "for_here"],
};

const pricingOptions: PricingOptions = {
  ingredientMarkup: 4.5,
  packagingMarkup: 1.5
};

const steamedMilkFromColdMilkForLatte = new Map<number, number>([
  [9, 7],
  [10.5, 8.5],
  [13, 9.5],
  [14.5, 11],
]);

const costForIngredient = new Map<string, number>([
  ["dairy_3_percent_cost_per_ounce", 0.05]
]);

const discounts = new Map<string, number>([["own_cup", 0.50]]);

const espressoCosting = {
  gramsPerShot: 18.5 / 2,
  ouncesPerShot: 1.5 / 2,
  costPerGram: 0.031
};

const database: CostingData = {
  costForPackaging,
  costForIngredient,
  packagingForHotDrinkSizeInOunces,
  customerOptions,
  pricingOptions,
  discounts,
  steamedMilkFromColdMilkForLatte,
  espressoCosting,
};

export default database;
