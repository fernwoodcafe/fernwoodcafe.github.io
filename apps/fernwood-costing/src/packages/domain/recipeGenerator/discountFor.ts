import type { RecipePermutation } from '.';
import type { CostingData } from './data/schema';

export default (
  recipe: Pick<RecipePermutation, "cupKind">,
  costingData: CostingData
) => {
  const discounts = [];

  if (recipe.cupKind === "own_cup") {
    discounts.push([
      recipe.cupKind, costingData.discounts.get("own_cup")
    ]);
  }

  const discountDollars = discounts.reduce((acc, next) => acc + next[1], 0);

  return {
    discounts,
    discountDollars
  };
};
