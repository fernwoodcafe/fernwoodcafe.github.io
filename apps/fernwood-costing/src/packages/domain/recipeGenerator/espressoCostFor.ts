import type { RecipePermutation } from '.';

export default (recipe: Pick<RecipePermutation, 'espressoShots'>) => ({
  espressoGrams: 18.5 * (recipe.espressoShots / 2),
  espressoFluidOunces: 1.5 * (recipe.espressoShots / 2),
  espressoCostDollars: 0.57 * (recipe.espressoShots / 2),
});
