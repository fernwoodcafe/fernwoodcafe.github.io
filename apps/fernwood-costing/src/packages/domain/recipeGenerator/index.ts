type AvailableCupKinds = "for_here" | "to_go" | "own_cup";
type AvailableCupSizesInOunces = 8 | 12 | 16;
type AvailableEspressoShots = 2 | 4 | 6;
type AvailableMilkAlternatives =
  | "dairy_one_percent"
  | "dairy_3_percent"
  | "dairy_10_percent"
  | "oat"
  | "almond";

export type CustomerOptions = {
  availableSizesInOunces: readonly AvailableCupSizesInOunces[];
  availableExpressoShots: readonly AvailableEspressoShots[];
  availableMilkAlternatives: readonly AvailableMilkAlternatives[];
  availableCups: readonly AvailableCupKinds[];
};

export type RecipeVariant = {
  cupSizeOunces: AvailableCupSizesInOunces;
  cupKind: AvailableCupKinds;
  espressoShots: AvailableEspressoShots;
  milkAlternative: AvailableMilkAlternatives;
};

const packagingCostFor = (
  options: Pick<RecipeVariant, "cupKind" | "cupSizeOunces">
) => {
  if (options.cupKind !== "to_go") {
    return 0;
  }

  // cup + lid + sleeve
  switch (options.cupSizeOunces) {
    case 8:
      // Note: We currently do not have 8 ounce sleeves;
      // people tend not to ask for them.
      return Math.round((0.13 + 0.12 + 0.00) * 100) / 100;
    case 12:
      return Math.round((0.17 + 0.13 + 0.08)* 100) / 100;
    case 16:
      return Math.round((0.19 + 0.13 + 0.08) * 100) / 100;
  }
};

export default (options: CustomerOptions): RecipeVariant[] =>
  options.availableSizesInOunces
    .map((cupSizeOunces) => ({
      cupSizeOunces,
    }))
    .flatMap((recipe) =>
      options.availableCups.map((cupKind) => ({
        ...recipe,
        cupKind,
      }))
    )
    .flatMap((recipe) =>
      options.availableExpressoShots.map((espressoShots) => ({
        ...recipe,
        espressoShots,
      }))
    )
    .flatMap((recipe) =>
      options.availableMilkAlternatives.map((milkAlternative) => ({
        ...recipe,
        milkAlternative,
      }))
    )
    .map((recipe) => ({
      ...recipe,
      // TODO [refactor] Clean up these formulas.
      espressoGrams: 18.5 * (recipe.espressoShots / 2),
      espressoFluidOunces: 1.5 * (recipe.espressoShots / 2),
      espressoCostDollars: 0.57 * (recipe.espressoShots / 2),
      packagingCostDollars: packagingCostFor(recipe),
    }));
