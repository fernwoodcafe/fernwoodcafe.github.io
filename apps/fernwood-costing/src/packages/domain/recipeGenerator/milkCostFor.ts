import type { RecipePermutation } from '.';
import { roundToTwoDecimalPlaces } from '../math/roundToDecimalPlaces';

export default (recipe: Pick<RecipePermutation, 'drinkSizeOunces' | 'espressoFluidOunces'>) => {
  const latteMilkSteamedToColdRatio = new Map<number, number>([
    [9, 7],
    [10.5, 8.5],
    [13, 9.5],
    [14.5, 11],
  ]);

  const dairy_3_percent_cost_per_ounce = 0.05;

  const milkSteamedOunces = recipe.drinkSizeOunces - recipe.espressoFluidOunces;
  const milkColdOunces = latteMilkSteamedToColdRatio.get(milkSteamedOunces);
  const milkCostDollars = roundToTwoDecimalPlaces(
    milkColdOunces * dairy_3_percent_cost_per_ounce
  );
  return {
    milkSteamedOunces,
    milkColdOunces,
    milkCostDollars,
  };
};
