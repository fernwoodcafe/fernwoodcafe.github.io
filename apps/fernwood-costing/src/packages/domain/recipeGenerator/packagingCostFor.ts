import { roundToTwoDecimalPlaces, type RecipePermutation } from '.';

export default (recipe: Pick<RecipePermutation, 'cupKind' | 'drinkSizeOunces'>) => {
  if (recipe.cupKind !== "to_go") {
    return {
      packagingCostDollars: 0,
    };
  }

  // cup + lid + sleeve
  switch (recipe.drinkSizeOunces) {
    case 8:
      // Note: We currently do not have 8 ounce sleeves;
      // people tend not to ask for them.
      return {
        packagingCostDollars: roundToTwoDecimalPlaces(0.13 + 0.12 + 0.0),
      };
    case 12:
      return {
        packagingCostDollars: roundToTwoDecimalPlaces(0.17 + 0.13 + 0.08),
      };
    case 16:
      return {
        packagingCostDollars: roundToTwoDecimalPlaces(0.19 + 0.13 + 0.08),
      };
  }
};

