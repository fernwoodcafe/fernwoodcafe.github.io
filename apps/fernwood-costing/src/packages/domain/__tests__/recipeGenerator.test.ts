import { beforeAll, describe, expect, it } from "vitest";
import recipeGenerator, {
  type RecipePermutation
} from "../recipeGenerator";
import database from '../recipeGenerator/data/database';
import dumpRecipes from './dumpRecipes';

const someExpectedPrices = [
  ['12 oz 2 shot dairy_3_percent for_here', 4.74]
];

// Arranage
describe("recipeGenerator", () => {
  let recipes: RecipePermutation[];
  beforeAll(() => {
    // Act
    recipes = recipeGenerator(database);

    // Dump
    dumpRecipes(recipes);
  });

  someExpectedPrices.forEach(([name, price]) => {
    it(`prices the ${name} at ${price}`, () => {
      const recipe = recipes.find(r => r.descriptiveName === name);
      expect(recipe.suggestedPrice).toEqual(price);
    }
    )
  });

  it("matches snapshot", () => {
    // Assert
    expect(recipes).toMatchSnapshot();
  });
});
