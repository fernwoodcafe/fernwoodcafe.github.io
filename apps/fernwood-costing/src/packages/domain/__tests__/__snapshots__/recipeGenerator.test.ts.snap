// Vitest Snapshot v1

exports[`recipeGenerator > matches snapshot 1`] = `
[
  {
    "cupKind": "to_go",
    "descriptiveName": "12 oz 2 shot dairy_3_percent to_go",
    "discountDollars": 0,
    "discounts": [],
    "drinkSizeOunces": 12,
    "espressoCostDollars": 0.57,
    "espressoFluidOunces": 1.5,
    "espressoGrams": 18.5,
    "espressoShots": 2,
    "ingredientCostDollars": 1,
    "ingredientMarkup": 4.75,
    "milkAlternative": "dairy_3_percent",
    "milkColdOunces": 8.5,
    "milkCostDollars": 0.43,
    "milkSteamedOunces": 10.5,
    "packaging": [
      "cup @ 12 oz",
      "lid @ 10 to 20 oz",
      "sleeve @ 12 to 20 oz",
    ],
    "packagingCostDollars": 0.38,
    "packagingMarkup": 1.5,
    "suggestedIngredientsPrice": 4.74,
    "suggestedPackagingPrice": 0.57,
    "suggestedPrice": 5.31,
    "totalCostDollars": 1.38,
  },
  {
    "cupKind": "own_cup",
    "descriptiveName": "12 oz 2 shot dairy_3_percent own_cup",
    "discountDollars": 0.5,
    "discounts": [
      [
        "own_cup",
        0.5,
      ],
    ],
    "drinkSizeOunces": 12,
    "espressoCostDollars": 0.57,
    "espressoFluidOunces": 1.5,
    "espressoGrams": 18.5,
    "espressoShots": 2,
    "ingredientCostDollars": 1,
    "ingredientMarkup": 4.75,
    "milkAlternative": "dairy_3_percent",
    "milkColdOunces": 8.5,
    "milkCostDollars": 0.43,
    "milkSteamedOunces": 10.5,
    "packaging": [],
    "packagingCostDollars": 0,
    "packagingMarkup": 1.5,
    "suggestedIngredientsPrice": 4.74,
    "suggestedPackagingPrice": 0,
    "suggestedPrice": 4.24,
    "totalCostDollars": 1,
  },
  {
    "cupKind": "for_here",
    "descriptiveName": "12 oz 2 shot dairy_3_percent for_here",
    "discountDollars": 0,
    "discounts": [],
    "drinkSizeOunces": 12,
    "espressoCostDollars": 0.57,
    "espressoFluidOunces": 1.5,
    "espressoGrams": 18.5,
    "espressoShots": 2,
    "ingredientCostDollars": 1,
    "ingredientMarkup": 4.75,
    "milkAlternative": "dairy_3_percent",
    "milkColdOunces": 8.5,
    "milkCostDollars": 0.43,
    "milkSteamedOunces": 10.5,
    "packaging": [],
    "packagingCostDollars": 0,
    "packagingMarkup": 1.5,
    "suggestedIngredientsPrice": 4.74,
    "suggestedPackagingPrice": 0,
    "suggestedPrice": 4.74,
    "totalCostDollars": 1,
  },
  {
    "cupKind": "to_go",
    "descriptiveName": "16 oz 2 shot dairy_3_percent to_go",
    "discountDollars": 0,
    "discounts": [],
    "drinkSizeOunces": 16,
    "espressoCostDollars": 0.57,
    "espressoFluidOunces": 1.5,
    "espressoGrams": 18.5,
    "espressoShots": 2,
    "ingredientCostDollars": 1.12,
    "ingredientMarkup": 4.75,
    "milkAlternative": "dairy_3_percent",
    "milkColdOunces": 11,
    "milkCostDollars": 0.55,
    "milkSteamedOunces": 14.5,
    "packaging": [
      "cup @ 16 oz",
      "lid @ 10 to 20 oz",
      "sleeve @ 12 to 20 oz",
    ],
    "packagingCostDollars": 0.4,
    "packagingMarkup": 1.5,
    "suggestedIngredientsPrice": 5.34,
    "suggestedPackagingPrice": 0.6,
    "suggestedPrice": 5.94,
    "totalCostDollars": 1.52,
  },
  {
    "cupKind": "own_cup",
    "descriptiveName": "16 oz 2 shot dairy_3_percent own_cup",
    "discountDollars": 0.5,
    "discounts": [
      [
        "own_cup",
        0.5,
      ],
    ],
    "drinkSizeOunces": 16,
    "espressoCostDollars": 0.57,
    "espressoFluidOunces": 1.5,
    "espressoGrams": 18.5,
    "espressoShots": 2,
    "ingredientCostDollars": 1.12,
    "ingredientMarkup": 4.75,
    "milkAlternative": "dairy_3_percent",
    "milkColdOunces": 11,
    "milkCostDollars": 0.55,
    "milkSteamedOunces": 14.5,
    "packaging": [],
    "packagingCostDollars": 0,
    "packagingMarkup": 1.5,
    "suggestedIngredientsPrice": 5.34,
    "suggestedPackagingPrice": 0,
    "suggestedPrice": 4.84,
    "totalCostDollars": 1.12,
  },
  {
    "cupKind": "for_here",
    "descriptiveName": "16 oz 2 shot dairy_3_percent for_here",
    "discountDollars": 0,
    "discounts": [],
    "drinkSizeOunces": 16,
    "espressoCostDollars": 0.57,
    "espressoFluidOunces": 1.5,
    "espressoGrams": 18.5,
    "espressoShots": 2,
    "ingredientCostDollars": 1.12,
    "ingredientMarkup": 4.75,
    "milkAlternative": "dairy_3_percent",
    "milkColdOunces": 11,
    "milkCostDollars": 0.55,
    "milkSteamedOunces": 14.5,
    "packaging": [],
    "packagingCostDollars": 0,
    "packagingMarkup": 1.5,
    "suggestedIngredientsPrice": 5.34,
    "suggestedPackagingPrice": 0,
    "suggestedPrice": 5.34,
    "totalCostDollars": 1.12,
  },
]
`;
